pkg/providers/aws/provider_redis.go

```diff
@@ -0,0 +1,280 @@
+}
+
+// DeleteStorage Delete elasticache replication group
+func (p *AWSRedisProvider) DeleteRedis(ctx context.Context, r *v1alpha1.Redis) error {
+	// resolve redis information for redis created by provider
+	redisConfig, stratCfg, err := p.getRedisConfig(ctx, r)
+	if err != nil {
+		return errorUtil.Wrapf(err, "failed to retrieve aws redis config for instance %s", r.Name)
+	}
+	if redisConfig.ReplicationGroupId == nil {
+		redisConfig.ReplicationGroupId = aws.String(r.Name)
+	}
+
+	// get provider aws creds so the redis cluster can be deleted
+	providerCreds, err := p.CredentialManager.ReconcileProviderCredentials(ctx, r.Namespace)
+	if err != nil {
+		return errorUtil.Wrap(err, "failed to reconcile aws provider credentials")
+	}
+
+	// setup aws redis cluster sdk session
+	cacheSvc := createCacheService(stratCfg, providerCreds)
+
+	// delete the redis cluster
+	return p.deleteRedisCluster(cacheSvc, redisConfig, ctx, r)
+}
+
+func (p *AWSRedisProvider) deleteRedisCluster(cacheSvc elasticacheiface.ElastiCacheAPI, redisConfig *elasticache.CreateReplicationGroupInput, ctx context.Context, r *v1alpha1.Redis) error {
+	// the aws access key can sometimes still not be registered in aws on first try, so loop
+	rgs, err := getReplicationGroups(cacheSvc)
+	if err != nil {
+		return err
+	}
+
+	// check if the cluster has already been deleted
+	var foundCache *elasticache.ReplicationGroup
+	for _, c := range rgs {
+		if *c.ReplicationGroupId == *redisConfig.ReplicationGroupId {
+			foundCache = c
+			break
+		}
+	}
+	// check if replication group does not exist and delete finalizer
+	if foundCache == nil {
+		// remove the finalizer added by the provider
+		resources.RemoveFinalizer(&r.ObjectMeta, DefaultFinalizer)
+		if err := p.Client.Update(ctx, r); err != nil {
+			return errorUtil.Wrapf(err, "failed to update instance as part of finalizer reconcile")
+		}
+		return nil
+	}
+	// check if replication group exists and is available
+	if *foundCache.Status == "available" {
+		// delete the redis cluster that was created by the provider
+		_, err = cacheSvc.DeleteReplicationGroup(&elasticache.DeleteReplicationGroupInput{
+			ReplicationGroupId:   redisConfig.ReplicationGroupId,
+			RetainPrimaryCluster: aws.Bool(false),
+		})
+		redisErr, isAwsErr := err.(awserr.Error)
+		if err != nil && !isAwsErr {
+			return errorUtil.Wrapf(err, "failed to delete elasticache cluster %s", *redisConfig.ReplicationGroupId)
+		}
+		if err != nil && isAwsErr {
+			if redisErr.Code() != elasticache.ErrCodeReplicationGroupNotFoundFault {
+				return errorUtil.Wrapf(err, "failed to delete elasticache cluster %s, aws error", *redisConfig.ReplicationGroupId)
+			}
+		}
+	}
+	return nil
+}
+
+// poll for replication groups
+func getReplicationGroups(cacheSvc elasticacheiface.ElastiCacheAPI) ([]*elasticache.ReplicationGroup, error) {
+	var rgs []*elasticache.ReplicationGroup
+	err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (done bool, err error) {
```

Comment:
If we fail here will it end up a requeue?

Reply:
No we don't, I'll change this